Fix bug when adding tier price to simple product from product edit page.

markbeazley opened the pull request:
Discovered a bug when adding/editing tier prices on a simple product as the form has a different root object Product vs Product Variant, I've managed to work around this, by checking if the form root is of type ProductInterface and if so grab the first ProductVariant and use that.

markbeazley commented:
Should I open an issue for this or would it be possible to get this considered for merging without an issue being created?

mamazu commented:
@kortwotze will have a look at it as he is currently using this in our production environment.

kortwotze commented:
Can you elaborate a bit more on what happened when you edited a simple product?

kortwotze commented on src/Validator/TierPriceUniqueValidator.php:
```diff
@@ -59,7 +61,15 @@ public function validate($value, Constraint $constraint): void
         }
 
-        $otherTierPrices = $this->context->getRoot()->getData()->getTierPrices();
+        $formData = $this->context->getRoot()->getData();
```
I'd suggest naming this `$productVariant` for clarity.

markbeazley replied:
I can change it to that, however there is a possibility it is a Product, and on the next line I call `$formData->hasVariants()` and `$productVariant->hasVariants()` seems a little odd.

kortwotze replied:
I see.

markbeazley commented:
Basically when you edit a tier price from the product edit route (rather than the variant edit route) it would try and call getTierPrices() on the Product rather than the variant due to the way the forms are nested.

kortwotze commented:
Mh - I can't reproduce it on my local sylius. Adding a tierprice on `admin/products/11/edit` is no problem here. Gets persisted in database and shown after reload.
Not trying to criticize here, just trying to comprehend properly. :)

kortwotze commented:
Disclaimer: I am using 1.5.0. where this file is not even part of. Okay, I take back of most that I said and try to understand what @mamazu has merged in 1.6.*.

markbeazley commented:
Yeah it is a bug in part of the validator I added when I added the customer groups features as annoyingly the built in UniqueEntity validator doesn't catch cases where a user tries to add two identical tier prices at the same time (it only checks against entities that were in the database at the time the form was submitted). So I wrote something that gets the submitted data and checks uniqueness in that data.

kortwotze commented:
Right - thanks for the insight. Now I am just wondering why the form can be rendered with a product as a root, rather than a variant (which should be the case all the time from what I see).

I still consider this fix valid though, just waiting for the full clarification and then we merge it. No issue needed here, I think :)

markbeazley commented:
Yeah I was a little puzzled by that myself, Not sure if I should have been using something else other than getRoot().

pierre-H commented:
same problem ...

kortwotze commented on src/Validator/TierPriceUniqueValidator.php:
```diff
@@ -59,7 +61,15 @@ public function validate($value, Constraint $constraint): void
-        $otherTierPrices = $this->context->getRoot()->getData()->getTierPrices();
+        $formData = $this->context->getRoot()->getData();
+        if ($formData instanceof ProductInterface && $formData->hasVariants()) {
+            $formData = $formData->getVariants()->first();
```
Maybe also check if its exactly 1 variant on the product, not just using the first.

markbeazley replied:
So change the if to this:

`if ($formData instanceof ProductInterface && count($formData->getVariants()) === 1)`

mamazu replied:
Looks good to me.

markbeazley replied:
I've made the change as mentioned above.

peterukena replied:
:+1:

mamazu replied:
Should we merge it?
